### PR TITLE
Implement native messenger interaction with the X selection buffer

### PR DESCRIPTION
**Pull request by glacambre**
I think I managed to get native_background.ts:clipboard() right but it should be thoroughly reviewed anyway, if I left a bug in there an attacker could potentially execute arbitrary programs on a user's computer when they copy something.

This fixes https://github.com/cmcaine/tridactyl/issues/76 .

### CODE REVIEW

**Comment by bovine3dom** [comment]
I like the idea that issue #666 could be the one that introduces an awful security hole into Tridactyl.

We'll check it over thoroughly : )

File: src/excmds.ts [hunk 1]
```diff
@@ -1912,35 +1971,32 @@ export async function clipboard(excmd: "open" | "yank" | "yankshort" | "yankcano
                 urls = await geturlsforlinks("rev", "canonical")
             }
             if (urls.length > 0) {
-                messageActiveTab("commandline_frame", "setClipboard", [urls[0]])
+                setclip(await urls[0])
```
Comment (glacambre) [inline src/excmds.ts]: No need for an await here

File: src/native_background.ts [hunk 2]
```diff
@@ -282,6 +282,23 @@ export async function getenv(variable: string) {
     return (await sendNativeMsg("env", { var: variable })).content
 }
 
+export async function clipboard(action: "set" | "get", str: string): Promise<MessageResp> {
```
Comment (glacambre) [inline src/native_background.ts]: Needs documentation + probably a name that won't conflict with excmds.ts:clipboard()

File: src/native_background.ts [hunk 3]
```diff
@@ -282,6 +282,35 @@ export async function getenv(variable: string) {
...
+    if (action == "set")
+        clipcmd += " -i"
+    else
+        clipcmd += " -o"
+
+    // We need to quote the whole string otherwise shell injection is possible
+    // Quoting the whole string means we need to escape all quotes within the string
+    // In order to do this, we use $'', which makes the shell understand that a \' means an escaped ', not a literal \'
+    // Because backslashes are now considered as beginning escape sequences, we need to escape them too!
+    clipcmd = `echo -n $'${str.replace("\\", "\\\\").replace("'", "\\'")}' | ${clipcmd}`
```
Comment (glacambre) [inline src/native_background.ts]: Just realized that $'' is bash-specific. I'll change this to heredocs + escaped backslashes and escaped dollars soon.

**Comment by glacambre** [comment]
I just realized that we might not need the native messenger at all. Maybe just creating an input field, filling it with the desired value and selecting its content with selectionStart/selectionEnd would work. I'll look into it.

File: src/native_background.ts [hunk 4]
```diff
@@ -282,6 +282,50 @@ export async function getenv(variable: string) {
...
+    // We're going to need to insert str, which we can't trust, in the clipcmd
+    // In order to do this safely we'll use here documents:
+    // http://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_07_04
+
+    // Find a delimiter that isn't in str
+    let heredoc = "TRIDACTYL"
+    while (str.search(heredoc) != -1) heredoc += Math.round(Math.random() * 10)
+
+    // Use delimiter to insert str into clipcmd's stdin
+    clipcmd = `${clipcmd} -i <<'${heredoc}'\n${str}\n${heredoc}\n`
```
Comment (bovine3dom) [inline src/native_background.ts]: @cmcaine please give your opinion on whether this is enough sanitisation. It looks plausible to me.

**Comment by bovine3dom** [comment]
This seems to work well, thanks @glacambre.

I am not confident enough in the security side of things to make any meaningful comments, so I'll bikeshed instead:

Could you add Vim-style commands to copy/"paste" from a specific clipboard, i.e,`"{+.*}{p,y}`?

File: src/native_background.ts (hunk 4 above)
Comment (cmcaine) [reply src/native_background.ts]: Looks good to me.

File: src/excmds.ts [hunk 5]
```diff
@@ -1883,6 +1883,60 @@ export async function get_current_url() {
...
+    // Functions to avoid retyping everything everywhere
+    let s = () => Native.clipboard("set", str)
+    let c = () => messageActiveTab("commandline_frame", "setClipboard", [str])
+
+    if (selection === true) return s()
+    if (selection === false) return c()
+
+    // "selection" is undefined here so let's try to determine what the user wants
+    let external = await config.getAsync("externalclipboard")
+    let promises = []
```
Comment (cmcaine) [inline src/excmds.ts]: don't need to name `external`

File: src/excmds.ts [hunk 6]
```diff
@@ -1912,35 +1970,34 @@ export async function clipboard(excmd: "open" | "yank" | "yankshort" | "yankcano
...
-            messageActiveTab("commandline_frame", "setClipboard", [content])
+            setclip((await activeTab()).title)
             break
         case "yankmd":
             content = "[" + (await activeTab()).title + "](" + (await activeTab()).url + ")"
-            messageActiveTab("commandline_frame", "setClipboard", [content])
+            setclip(content)
             break
         case "open":
-            await messageActiveTab("commandline_content", "focus")
```
Comment (cmcaine) [inline src/excmds.ts]: I presume you've checked that these aren't required? I think I was probably being overzealous when I put them in.

File: src/config.ts [hunk 7]
```diff
@@ -286,6 +286,8 @@ const DEFAULTS = o({
     //          (but we are probably happy to add your terminal to the list if it isn't already there).
     editorcmd: "auto",
     browser: "firefox",
+    externalclipboard: "last", // "only" | "first" | "last" | "never", see excmds.ts:clipboard() to know what this does
```
[the remark this answers is not in the document]
Comment (glacambre) [reply src/config.ts]: What do you prefer, keeping `externalclipboard` or switching to `yankto/putfrom`? I think we shouldn't care too much about users shooting themselves in the foot.

**Comment by glacambre** [comment]
Rebased on master.

File: src/config.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (bovine3dom) [reply src/config.ts]: I prefer yankto and putfrom :)

File: src/excmds.ts [hunk 8]
```diff
@@ -2077,57 +2120,63 @@ export async function yank(...content: string[]) {
 
     If `excmd == "yankmd"`, copy the title and url of the open page formatted in Markdown for easy use on sites such as reddit.
 
-    Unfortunately, javascript can only give us the `clipboard` clipboard, not e.g. the X selection clipboard.
+    If you're on Linux and the native messenger is installed, Tridactyl will call an external binary (either xclip or xsel) to read or write to your X selection buffer.
+
+    When doing a read operation (i.e. open or tabopen), if "externalclipboard" is set to "only" or "first", the X selection buffer will be read instead of the clipboard. Set "externalclipboard" to "last" or "never" to use the clipboard.
+
+    When doing a write operation, if "externalclipboard" is set to "only", only the X selection buffer will be written to. If "externalclipboard" is set to "first" or "last", both the X selection and the clipboard will be written to. If "externalclipboard" is set to "never", only the clipboard will be written to.
```
Comment (bovine3dom) [inline src/excmds.ts]: Could you document externalclipboardcmd and the flags it must support? It's plausible that some people might want to use a complicated clipboard manager.

**Pull request merged**